dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusPTX.java

```diff
@@ -0,0 +1,310 @@
+                if (ctx.scope() != null) {
+                    throw new ParsingException("Weak load instruction doesn't need scope: " + ctx.scope().content);
+                }
+                scope = Tag.PTX.SYS;
+                break;
+            case Tag.PTX.ACQ:
+            case Tag.PTX.RLX:
+                scope = ctx.scope().content;
+                break;
+            default:
+                throw new ParsingException("Load instruction doesn't support sem: " + mo);
+        }
+        Load load = EventFactory.PTX.newTaggedLoad(register, location, mo, scope);
+        load.addFilters(ctx.load().loadProxy);
+        return programBuilder.addChild(mainThread, load);
+    }
+
+    @Override
+    public Object visitAtomConstant(LitmusPTXParser.AtomConstantContext ctx) {
+        Register register_destination = programBuilder.getOrCreateRegister(mainThread, ctx.register().getText(), getArchPrecision());
+        MemoryObject object = programBuilder.getOrNewObject(ctx.location().getText());
+        IValue constant = new IValue(new BigInteger(ctx.constant().getText()), getArchPrecision());
+        IOpBin op = ctx.operation().op;
+        String mo = ctx.sem().content;
+        String scope;
+        if (mo.equals(Tag.PTX.ACQ_REL) || mo.equals(Tag.PTX.RLX)) {
+            scope = ctx.scope().content;
+        } else {
+            throw new ParsingException("Atom instruction doesn't support sem: " + mo);
+        }
+        AtomOp atom = EventFactory.PTX.newTaggedAtomOp(object, register_destination, constant, op, mo, scope);
+        atom.addFilters(ctx.atom().atomProxy);
+        return programBuilder.addChild(mainThread, atom);
+    }
+
+    @Override
+    public Object visitAtomRegister(LitmusPTXParser.AtomRegisterContext ctx) {
+        Register register_destination = programBuilder.getOrCreateRegister(mainThread, ctx.register().get(0).getText(), getArchPrecision());
+        MemoryObject object = programBuilder.getOrNewObject(ctx.location().getText());
+        Register register_operand = programBuilder.getOrCreateRegister(mainThread, ctx.register().get(1).getText(), getArchPrecision());
+        IOpBin op = ctx.operation().op;
+        String mo = ctx.sem().content;
+        String scope;
+        if (mo.equals(Tag.PTX.ACQ_REL) || mo.equals(Tag.PTX.RLX)) {
+            scope = ctx.scope().content;
+        } else {
+            throw new ParsingException("Atom instruction doesn't support sem: " + mo);
+        }
+        AtomOp atom = EventFactory.PTX.newTaggedAtomOp(object, register_destination, register_operand, op, mo, scope);
+        atom.addFilters(ctx.atom().atomProxy);
+        return programBuilder.addChild(mainThread, atom);
+    }
+
+    @Override
+    public Object visitRedConstant(LitmusPTXParser.RedConstantContext ctx) {
+        MemoryObject object = programBuilder.getOrNewObject(ctx.location().getText());
+        IValue constant = new IValue(new BigInteger(ctx.constant().getText()), getArchPrecision());
```

Comment:
this can probably be replaced by `(IConst) ctx.constant().accept(this)`

Please make the same change everywhere

Reply:
Replaced